lsp: Implement support for the `textDocument/diagnostic` command

vitallium opened the pull request:
Closes [#13107](https://github.com/zed-industries/zed/issues/13107)

Enabled pull diagnostics by default, for the language servers that declare support in the corresponding capabilities.

```
"diagnostics": {
    "lsp_pull_diagnostics": {
        "enabled": false
    }
}
```
settings can be used to disable the pulling.

Release Notes:

- Added support for the LSP `textDocument/diagnostic` command.

# Brief

This is draft PR that implements the LSP `textDocument/diagnostic` command. The goal is to receive your feedback and establish further steps towards fully implementing this command. I tried to re-use existing method and structures to ensure:

1. The existing functionality works as before
2. There is no interference between the diagnostics sent by a server and the diagnostics requested by a client.

The current implementation is done via a new LSP command `GetDocumentDiagnostics` that is sent when a buffer is saved and when a buffer is edited. There is a new method called `pull_diagnostic` that is called for such events. It has debounce to ensure we don't spam a server with commands every time the buffer is edited. Probably, we don't need the debounce when the buffer is saved.

All in all, the goal is basically to get your feedback and ensure I am on the right track. Thanks!


## References

1. https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_pullDiagnostics

## In action

You can clone any Ruby repo since the `ruby-lsp` supports the pull diagnostics only.

Steps to reproduce:

1. Clone this repo https://github.com/vitallium/stimulus-lsp-error-zed
2. Install Ruby (via `asdf` or `mise).
4. Install Ruby gems via `bundle install`
5. Install Ruby LSP with `gem install ruby-lsp`
6. Check out this PR and build Zed
7. Open any file and start editing to see diagnostics in realtime.


https://github.com/user-attachments/assets/0ef6ec41-e4fa-4539-8f2c-6be0d8be4129

vitallium committed:
lsp: Implement support for the `textDocument/diagnostic` command

vitallium commented on crates/language/src/language.rs:
```diff
@@ -201,6 +201,7 @@ pub struct CachedLspAdapter {
     pub name: LanguageServerName,
     pub disk_based_diagnostic_sources: Vec<String>,
     pub disk_based_diagnostics_progress_token: Option<String>,
+    pub previous_document_diagnostic_result_id: Option<String>,
```
Still not sure about the place to store this token that Zed should store and use between sending/receiving `textDocument/diagnostic` commands.

vitallium replied:
Moved the `previous_document_diagnostic_result_id` to `LanguageServerState` enum. Let me know if that's ok.

vitallium commented on crates/lsp/src/lsp.rs:
```diff
@@ -696,6 +696,10 @@ impl LanguageServer {
                     }),
+                    diagnostic: Some(DiagnosticClientCapabilities {
+                        dynamic_registration: Some(true),
+                        related_document_support: Some(true),
```
Alternatively, we could specify that Zed does not support handling diagnostics for related documents currently with `Some(false)`.

vitallium replied:
>What's the difference from Zed's perspective from setting this to false or true?

The difference is, with `true`, Zed tells a server that it can process diagnostics for related documents when it receives diagnostics.

>For the related documents, we would get the diagnostics in the same format, but for different files and with different metadata?

Yes, correct.

>If that's it, we seem to be able to display these too, and can support this capability?

Pretty much, yes. I will implement that feature.

vitallium commented on crates/project/src/lsp_store.rs:
```diff
@@ -2886,6 +2887,15 @@ impl LspStore {
                 .log_err();
+
+            let buffer_handle = buffer_handle.clone();
+            cx.spawn(move |this, mut cx| async move {
+                this.update(&mut cx, |this, cx| {
```
I have doubts about these lines. Am I doing this correctly?

vitallium replied:
Changed that to `MultiLspQuery`.

vitallium commented on crates/project/src/lsp_store.rs:
```diff
@@ -2939,11 +2949,58 @@ impl LspStore {
+                    language_server_id,
+                    lsp::PublishDiagnosticsParams {
+                        uri: uri.clone(),
+                        diagnostics: diagnostics.unwrap(),
```
- [ ] Remove `unwrap`

vitallium commented:
Whoops, conflicts. I will resolve them in a bit.

vitallium commented:
> Whoops, conflicts. I will resolve them in a bit.

Done.

vitallium commented on crates/project/src/lsp_command.rs:
```diff
@@ -181,6 +181,11 @@ pub(crate) struct LinkedEditingRange {
 }
 
+#[derive(Debug)]
+pub(crate) struct GetDocumentDiagnostics {
```
Is implementing such a command via `LspCommand` right? The current method for getting diagnostics is done directly in the `LspStore`, but I noticed there is some work for moving code to support SSH. So, this is an open question.

vitallium committed:
lsp: store and use previous result of document diagnostics

vitallium committed:
lsp: Remove the `version` field from GetDocumentDiagnostics

vitallium committed:
lsp: switch to `DiagnosticEntry` for pull diagnostics

vitallium commented on crates/project/src/lsp_command.rs:
```diff
@@ -3020,3 +3028,179 @@ impl LspCommand for LinkedEditingRange {
     }
 }
+
+fn lsp_diagnostic_to_diagnostic(diagnostic: lsp::Diagnostic) -> Diagnostic {
```
I am thinking about moving this to `language.rs`

vitallium replied:
Pull-based diagnostics go through `LspStore` methods to fill these properties correctly now.

vitallium commented:
Ok, this is ready for the first round of review. Thanks!

vitallium commented:
Resolved merge conflicts.

SomeoneToIgnore reviewed (commented):
One thing I do not understand is how both diagnostics pull and diagnostics push models are supposed to live together.

* For instance, rust-analyzer has its own diagnostics, fast to update, and the cargo check-based ones, which may take seconds to update on large projects.
If we start to wait for all diagnostics or nothing, it might get slower — right now it's ok, as r-a pushes slow and fast ones separately.
How does this feature work on Zed?

* How should the LSP server and the client agree what to use?
It seems, that the push method is the default, and pull is an opt-in one, that will be used right after we declare this new capability?

* Is it ok to handle both push and pull at the same time, as we do now?
We have to include potential refresh requests that we either have to support (as declared currently in the capabilities) or declare incapable.

vitallium commented:
@SomeoneToIgnore Thanks for this amazing review! This is what I've been waiting for. I will start working on your comments and answering your questions about handling both push and pull models for diagnostics. Do you want me to convert this PR to draft to mark it as non-ready?

vitallium committed:
lsp: use task-based approach for pulling diagnostics

lsp: use task-based approach for pulling diagnostics

vitallium committed:
lsp: make diagnostics results optional

Requesting diagnostics for a document
can return an empty result meaning
there are no new diagnostics for the current document.
Returning an empty vector could indicate
that. It would be good to avoid that
and return `None` if diagnostics
haven't changed.

vitallium commented:
# 2024-11-09 Update

With the great review from @SomeoneToIgnore, I've managed to utilize the `MultiLspQuery` approach instead of querying the primary language server.  
In its current state, I can see Zed sends `textDocument/diagnostics` requests, but it does not display them yet.

One major blocker is the RPC implementation, which I am struggling with. There are two different structures for diagnostics: `lsp::Diagnostic` and `Diagnostic`. Despite being similar, they are different. The `Diagnostic` structure has fields like `group_id` and `is_disk_based` that do not exist in the `lsp::Diagnostic` structure, which is used for sending and receiving LSP commands. The RPC implementation requires serializing and deserializing diagnostics with `protobuf`, and I do not know how to achieve that without adding messages to the protocol for `lsp::Diagnostic`.

There is a method, `update_diagnostics`, in the `LspStore` that fills fields in the `Diagnostic` structure. I will check if I can use it. Additionally, this PR is missing pieces for the dynamic registration of LSP capabilities for pulling diagnostics. I also suspect there are missing pieces for the RPC stuff. I will have a look into both things too.

**TODO list:**

- Finish support for the RPC.
- Check dynamic registrations.
- Check how to use the `previous_result_id` that an LSP server sends back to indicate diagnostics haven't changed since the last request.
- Implement handling of diagnostics for relevant documents. I think the easiest way is to create a new method like `handle_relative_documents_diagnostics` in the `LspStore` and call it from the `response_from_lsp` method in `GetDocumentDiagnostics`.
- Check [the `refresh_support` thing](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnostic_refresh).

Thanks!

vitallium committed:
lsp: add struct for LSP pull-based diagnostics

SomeoneToIgnore commented:
To respond to the RPC part, it's always so that RPC clients have no access to LSP, it runs on the host machine only (either an RPC owner, or on a headless ssh part).

So, we always pass the LSP data over the wire and derive the additional properties on the client, hence you can (and should) use existing methods that fill those properties.
If that's not the case, let's have a look together at some point — ping me when more attention is needed to this PR.

vitallium committed:
lsp: improve task for pulling document diagnostics

vitallium committed:
lsp: make LspDiagnostics response optional

vitallium commented:
There have been a lot of changes recently in the related crates. I will double-check all the changes here. Also, I may need to review the implementation approach in [this commit](https://github.com/zed-industries/zed/pull/19230/commits/9ddc70ce0c634a78127a7ca328be48690e5b15c3)

The reason why I went this way is that the `LspStore::update_diagnostics` function accepts `lsp::PublishDiagnosticsParams` as one of its parameters. Diagnostics pulled via `textDocument/diagnostic` are not fully compatible with `PublishDiagnosticsParams`; for instance, there is no `uri: Url` field. So, I have had to introduce separate structs for pulled diagnostics.

vitallium committed:
lsp: handle `related_information` and `tags`

vitallium committed:
lsp: handle pull-based diagnostics model for related documents

vitallium commented on crates/lsp/src/lsp.rs:
```diff
@@ -577,7 +577,7 @@ impl LanguageServer {
                     }),
                     diagnostic: Some(DiagnosticWorkspaceClientCapabilities {
-                        refresh_support: None,
+                        refresh_support: Some(true),
```
Disabled it for now and will be implemented later.

vitallium replied:
Whoops, that was the comment for `dynaimc_registration`. I will check how it's done for inlay hints and implement for pull-based diagnostics.

vitallium replied:
Added support for `workspace/diagnostic/refresh`. I used `rust-analyzer` to test that. Here is the corresponding log when I opened the `zed-ruby` project and modified the Cargo.toml file to trigger workspace configuration change and recalculation of all diagnostics in a workspace:

```jsonc
// Receive:
{"jsonrpc":"2.0","id":45,"method":"workspace/diagnostic/refresh"}
// Receive:
{"jsonrpc":"2.0","method":"$/progress","params":{"token":"rustAnalyzer/Indexing","value":{"kind":"report","cancellable":false,"percentage":23}}}
// Omitted progress messages
// Send:
{"jsonrpc":"2.0","id":7,"method":"textDocument/diagnostic","params":{"textDocument":{"uri":"file:///Users/vslobodin/Development/zed-ruby/src/ruby.rs"},"identifier":null,"previousResultId":null}}
// Receive:
{"jsonrpc":"2.0","id":7,"result":{"kind":"full","resultId":"rust-analyzer","items":[]}}
```

vitallium commented on crates/lsp/src/lsp.rs:
```diff
@@ -696,6 +696,10 @@ impl LanguageServer {
                         ..TextDocumentSyncClientCapabilities::default()
                     }),
+                    diagnostic: Some(DiagnosticClientCapabilities {
+                        dynamic_registration: Some(true),
```
Disabled this capability for now, and will implement later.

vitallium commented on crates/project/src/lsp_command.rs:
```diff
@@ -3022,3 +3030,179 @@ impl LspCommand for LinkedEditingRange {
+        match message {
+            lsp::DocumentDiagnosticReportResult::Report(report) => match report {
+                lsp::DocumentDiagnosticReport::Full(report) => {
+                    // TODO: Handle related_documents in the report
```
I've added handling of diagnostics for related documents.

vitallium commented on crates/project/src/lsp_store.rs:
```diff
@@ -6468,6 +6530,7 @@ impl LspStore {
                     language: language.clone(),
                     server: language_server.clone(),
                     simulate_disk_based_diagnostics_completion: None,
+                    previous_document_diagnostic_result_id: None,
```
`previous_document_diagnostic_result_id` was removed for now.

vitallium commented on crates/project/src/lsp_store.rs:
```diff
@@ -3024,11 +3034,63 @@ impl LspStore {
 
         for language_server_id in self.language_server_ids_for_buffer(buffer.read(cx), cx) {
             self.simulate_disk_based_diagnostics_events_if_needed(language_server_id, cx);
+            self.pull_diagnostic(language_server_id, buffer.clone(), cx);
```
This was resolved with the `MultiLspQuery` implementation and task-based approach.

vitallium commented on crates/project/src/lsp_store.rs:
```diff
@@ -3024,11 +3034,63 @@ impl LspStore {
+                .unwrap_or_default();
+
+            buffer_handle.update(&mut cx, |buffer, cx| {
+                let set = DiagnosticSet::from_sorted_entries(diagnostics, &snapshot);
```
Good question! This was resolved with the MultiLspQuery implementation and task-based approach.

vitallium commented on crates/editor/src/editor.rs:
```diff
@@ -1157,6 +1160,10 @@ impl Editor {
                 project_subscriptions.push(cx.subscribe(project, |editor, _, event, cx| {
                     if let project::Event::RefreshInlayHints = event {
                         editor.refresh_inlay_hints(InlayHintRefreshReason::RefreshRequested, cx);
+                    } else if let project::Event::LanguageServerAdded(_, _, _) = event {
+                        editor.tasks_pull_diagnostics_task = Some(editor.refresh_diagnostics(cx));
+                    } else if let project::Event::LanguageServerRemoved(_) = event {
+                        editor.tasks_pull_diagnostics_task = Some(editor.refresh_diagnostics(cx));
```
**Question:** Not sure if triggering the task for pulling diagnostics here is the correct approach. My thinking is with the `MultiLspQuery` we want to pull new diagnostics sostics every time new LSP is added or removed.

vitallium commented:
@SomeoneToIgnore Hi! Can you please do a smoke check of this PR? Just to guide me if I am moving towards the right direction with the task-based approach and additional protobuf messages. The latter is the major question to me since I have had to do it to properly pass LSP diagnostics between Zed and LSPs. The main difference is the structure of [push-based](https://github.com/gluon-lang/lsp-types/blob/master/src/lib.rs#L2421) and [pull-based diagnostics](https://github.com/gluon-lang/lsp-types/blob/master/src/document_diagnostic.rs#L101) is different. Thanks for the suggestion about using `MultiLspQuery`, the implementation looks way better than before. I use `Ruby LSP` for testing because this is the only server with pull-based diagnostics that I can use currently. I know that `rust-analyzer` also supports it but I wasn't able to enable them.

Thanks again!

vitallium commented on crates/project/src/lsp_store.rs:
```diff
@@ -3024,11 +3034,63 @@ impl LspStore {
         None
     }
 
+    fn pull_diagnostic(
```
Yes, added a basic test for pull-based diagnostics.

SomeoneToIgnore commented:
Thank you for working on this.
I'll be able to look at this somewhere after 9th of Jan.

vitallium committed:
lsp: make LspDiagnostics response optional

vitallium committed:
lsp: impl `display_name` for GetDocumentDiagnostics LSP command

vitallium commented:
I see the following error message in the logs:

```
[2025-01-14T08:53:35+01:00 ERROR project::lsp_store] No LSP snapshots found for buffer with path "app/frontend/javascripts/parse_responses.js"
```

Not sure what that means exactly but I will have a look.

SomeoneToIgnore commented:
Sorry, still crawling towards your PR, which I remember about, but other tasks are always popping.

That error is 
https://github.com/zed-industries/zed/pull/22934

and not necessarily related to your changes, but it is something good to fix indeed.

vitallium commented:
> Sorry, still crawling towards your PR, which I remember about, but other tasks are always popping.

No worries! I understand how it goes. There's always a long list of tasks and priorities can shift. I appreciate you keeping my PR in mind. Thanks!

SomeoneToIgnore commented:
Just a small note about that 
`[2025-01-14T08:53:35+01:00 ERROR project::lsp_store] No LSP snapshots found for buffer with path "app/frontend/javascripts/parse_responses.js"` log message: it turned out to be overly pedantic and triggering for Zig and our settings.json out of the blue, so we're either forward-fixing this or reverting, so seems like nothing broken in the current PR.

vitallium commented:
> Just a small note about that `[2025-01-14T08:53:35+01:00 ERROR project::lsp_store] No LSP snapshots found for buffer with path "app/frontend/javascripts/parse_responses.js"` log message: it turned out to be overly pedantic and triggering for Zig and our settings.json out of the blue, so we're either forward-fixing this or reverting, so seems like nothing broken in the current PR.

Aha, good to know. Thank you!

vitallium committed:
lsp: rework a task for pulling diagnostics

vitallium committed:
lsp: remove position from `GetDocumentDiagnostic` message

vitallium commented on crates/editor/src/editor.rs:
```diff
@@ -1179,6 +1183,10 @@ impl Editor {
                 project_subscriptions.push(cx.subscribe(project, |editor, _, event, cx| {
                     if let project::Event::RefreshInlayHints = event {
                         editor.refresh_inlay_hints(InlayHintRefreshReason::RefreshRequested, cx);
+                    } else if let project::Event::LanguageServerAdded(_, _, _)
+                    | project::Event::LanguageServerRemoved(_) = event
+                    {
+                        editor.refresh_diagnostics(cx);
```
There is something I don't understand currently. Subscribing here to these events break the `test_move_to_enclosing_bracket` test. Surely I do know that some of my changes caused that but I just have no idea. Will continue my investigation.

vitallium replied:
Removing the debounce in `Editor::refresh_diagnostics makes tests green:

```
cx.background_executor()
  .timer(DOCUMENT_DIAGNOSTICS_DEBOUNCE_TIMEOUT)
  .await;
```

I guess I am blocking something there when I start pulling diagnostics upoon receiving `project::Event::LanguageServerAdded` or `LanguageServerRemoved`.

vitallium replied:
Simply waiting for the debounce makes the test green. Tho, I am not sure this is the good fix.

SomeoneToIgnore replied:
Yeah, this is very odd and somewhat concerning.
This fix could've been fine if the test failed on something like "something is not parked and we're blocked on waiting for LSP response" but as I see when running this test from the branch without the debounce, that we simply not move the caret.
So, the fix is not good and we need to dig deeper to understand why does not it move.

SomeoneToIgnore replied:
Just chatted about that as it surfaced on another branch: seems that we can use `cx.run_until_parked()` instead, as 

> the test replaces an editor's text but does not necessarily wait for the parsing to complete
> and then tries to use the syntax tree

So, not an issue of the current PR.
This will get fixed in another branch and gets merged, but we can change it here too, just to be on par with a proper fix.

vitallium replied:
Thanks! `cx.run_until_parked()` worked!

vitallium committed:
lsp: fix `test_move_to_enclosing_bracket` test

Starting pulling diagnostics upon receiving
`project::Event::LanguageServerAdded`
or `LanguageServerRemoved` results in a failed test
because pulling uses a debounce under the hood.
Simply waiting for debounce makes the test green.

vitallium committed:
lsp: add support for `workspace/diagnostic/refresh`

SomeoneToIgnore commented on crates/editor/src/editor_tests.rs:
```diff
@@ -15047,3 +15049,103 @@ fn assert_hunk_revert(
+            }
+        });
+
+    // Opening file should trigger diagnostics
```
I think this happens not because the file just is open, but because Zed initializes the language server and inserts it.
We might want to open another file of the same type later in the test to check this comment's statement.

SomeoneToIgnore commented on crates/project/src/lsp_command.rs:
```diff
@@ -4,14 +4,15 @@ use crate::{
 use clock::Global;
 use collections::HashSet;
-use futures::future;
+use futures::future::{self};
```
```suggestion
use futures::future;
```

Looks odd.

SomeoneToIgnore commented on crates/project/src/lsp_store.rs:
```diff
@@ -5850,6 +5940,47 @@ impl LspStore {
+                    .update(&mut cx, |project, cx| {
+                        project.request_multiple_lsp_locally(
+                            &buffer,
+                            Some(0),
```
That is odd, why not `None` so we can query all related LSP servers?

vitallium replied:
Changed that. Thanks!

SomeoneToIgnore commented on crates/project/src/lsp_store.rs:
```diff
@@ -4563,6 +4587,72 @@ impl LspStore {
+        } else {
+            let all_actions_task = self.request_multiple_lsp_locally(
+                &buffer_handle,
+                Some(0),
```
Same comment as somewhere around, we might want to use `None` here instead?

vitallium replied:
Changed that. Thanks!

SomeoneToIgnore commented on crates/project/src/lsp_command.rs:
```diff
@@ -3178,3 +3183,477 @@ impl LspCommand for LinkedEditingRange {
+    pub fn deserialize_lsp_diagnostic(diagnostic: proto::LspDiagnostic) -> Result<lsp::Diagnostic> {
+        let start = diagnostic
+            .start
+            .ok_or_else(|| anyhow!("invalid start range"))?;
```
```suggestion
            .context("invalid start range")?;
```

I know that there are a lot of places around doing old version, but just in case you want to change this to something more idiomatic here and below.

SomeoneToIgnore commented on crates/project/src/lsp_command.rs:
```diff
@@ -3178,3 +3183,477 @@ impl LspCommand for LinkedEditingRange {
+            Some(lsp::DiagnosticServerCapabilities::RegistrationOptions(options)) => {
+                options.diagnostic_options.identifier.clone()
+            }
+            None => None,
```
That seems somewhat wrong, as we're not supposed to send diagnostic requests when there's no corresponding capabilities?
Should we `.context(..)?` instead?

vitallium replied:
Indeed, thanks for noticing that!

vitallium replied:
That's interesting. By following your advice and adding `.context(..)?` here, I found that `identifier` is `None` for some reason:

```
Some(Options(DiagnosticOptions { identifier: None, inter_file_dependencies: true, workspace_diagnostics: false, work_done_progress_options: WorkDoneProgressOptions { work_done_progress: None } }))
```

But it should not be so. The identifier is `None` for `rust-analyzer` and `ruby-lsp`. That smells like an issue.

vitallium replied:
As per https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_diagnostic

>The additional identifier  provided during registration

I see that `rust-analyzer` does not set this field at all https://github.com/rust-lang/rust-analyzer/blob/master/crates/rust-analyzer/src/lsp/capabilities.rs#L168 So, I feel like we should not exit early if identifier is missing.

SomeoneToIgnore replied:
Nice catch.
r-a might plain wrong with this, but sure, let's be more lenient and just warn about this fact.

SomeoneToIgnore commented on crates/project/src/lsp_command.rs:
```diff
@@ -3178,3 +3183,477 @@ impl LspCommand for LinkedEditingRange {
+                                lsp::DocumentDiagnosticReportKind::Unchanged(_) => (),
+                            }
+                        }
+                    }).expect("Failed to update diagnostics for related documents");
```
I'm concerned by the amount of new `.unwrap()`, `.expect(..)` in the new code, esp. given that this function returns `Result`.
Let's not write the code that may panic here, as it's quite a core feature, queried frequently, and we're parsing things from a potentially malicious server.

vitallium replied:
Yes, totally makes sense. Changed that. The only ugly part left there is getting the URI of the document. We need that URI for `update_diagnostics` method on the `lsp_store`. If you have any suggestions for that part, please let me know. Thanks!

SomeoneToIgnore commented on crates/project/src/lsp_command.rs:
```diff
@@ -3178,3 +3183,477 @@ impl LspCommand for LinkedEditingRange {
 }
+
+impl GetDocumentDiagnostics {
+    pub fn deserialize_lsp_diagnostic(diagnostic: proto::LspDiagnostic) -> Result<lsp::Diagnostic> {
```
```suggestion
    fn deserialize_lsp_diagnostic(diagnostic: proto::LspDiagnostic) -> Result<lsp::Diagnostic> {
```

Same for serialization.
Is there a point in these impls being separate, or could they be moved entirely into the `impl LspCommand for GetDocumentDiagnostics`?

Also, somewhat surprised that we could not reuse the old code for that, we sure did get diagnostics pushed from LSP before, what is the issue with that?

vitallium replied:
>Same for serialization.

Done. Thanks!

>Is there a point in these impls being separate, or could they be moved entirely into the impl LspCommand for GetDocumentDiagnostics?

I'm not sure if I fully understand that. Don't we need to define them in the `LspCommand` trait first?

SomeoneToIgnore commented on crates/project/src/lsp_command.rs:
```diff
@@ -3178,3 +3183,477 @@ impl LspCommand for LinkedEditingRange {
+                None => None,
+            },
+            related_information: Some(related_information),
+            tags: Some(vec![]),
```
??

There's sure a `tags` field in the `proto` counterpart.

vitallium replied:
Whoops, sorry, my bad. Fixed!

SomeoneToIgnore commented on crates/project/src/lsp_command.rs:
```diff
@@ -3178,3 +3183,477 @@ impl LspCommand for LinkedEditingRange {
+                                            uri,
+                                            version: None,
+                                        },
+                                        &[],
```
How true is this?
There's a special code path for rust-analyzer's diagnostics that used to check whether the diagnostics pushed came from `cargo` or from r-a, is it so that we're losing this info now?

vitallium replied:
Hmm, yeah, you are correct. For Rust, there is `rustc` source - https://github.com/vitallium/zed/blob/17db3cd154e733edb666c468a273a69941b28ac1/crates/languages/src/rust.rs#L220 I don't think there is a way to get disk based sources inside the `lsp_command` so probably there should be another method in `lsp_store` that combines pulled diagnostics and disk-based sources. I will check that.

vitallium replied:
Actually, I found that it's possible to this:

```rust
let language_server_adapter = lsp_store
            .update(&mut cx, |lsp_store, _| {
                lsp_store.language_server_adapter_for_id(server_id)
            })?
            .ok_or_else(|| anyhow!("no such language server"))?;
````

in the `response_from_lsp` method to get LSP adapter which has the method `disk_based_diagnostic_sources` that is required here.

SomeoneToIgnore reviewed (commented):
Sorry for taking so long, now I'm back to business and hopefully able to respond to updates more quickly.

Main concerns I have for this PR:

* I want to preserve the existing diagnostics pushes, just in case certain servers are not able to do other things and just in case we mess up this impl, so people can use a fall back.
Also, I think I've raised it in previous reviews, we do not want both modes to work simultaneously?
I see, nothing is removed for pushing handle code, so that's possible?
Given all that, I expected somewhere in the language/language server-related logic, a piece of `if`-based logic to determine which path are we going with.

Another concern why I want to leave the old mechanism is the fact that we will loose "project diagnostics" functionality if we fully switch to this feature? 
As now, after changing one file, we need to query all files in the project to ensure we capture entire "all errors in the project" state, if the language server does not bother to send related files' diagnostics.

* I think, we do not query enough diagnostics, but maybe I'm wrong? Voiced that thought in the `editor.rs` comments.

* We still have some new code added for diagnostics [de]serialization, what was the issue with the existing one that handled pushes via LSP?

vitallium committed:
Fixes after #22632 landed

vitallium committed:
Avoid spamming the pull diagnostics task starts

vitallium commented on crates/editor/src/editor.rs:
```diff
@@ -10526,6 +10536,68 @@ impl Editor {
+        }
+
+        if self.tasks_pull_diagnostics_task.is_some() {
+            log::warn!(
```
Ah, yes, all of that makes sense!

I applied the following changes in 6fb9bc31015392ddb20d5c8804d5746dce80147e:

1. Changed the type of `tasks_pull_diagnostics_task` to `Task<()>` and initialization it with `Task::ready(())`.
2. Removed the `if self.tasks_pull_diagnostics_task.is_some() {` condition.

vitallium commented on crates/editor/src/editor.rs:
```diff
@@ -10526,6 +10536,68 @@ impl Editor {
+        let (cursor_buffer, _) = buffer.text_anchor_for_position(cursor_position.start, cx)?;
+        let (end_buffer, _) = buffer.text_anchor_for_position(cursor_position.end, cx)?;
+
+        if cursor_buffer != end_buffer {
```
I'm not entirely sure what you mean, but let me think it over for a bit.

SomeoneToIgnore replied:
Related to https://github.com/zed-industries/zed/pull/19230/files#r1922343682

To rephrase, every editor is backed by a multibuffer, and if those contain >1 buffers (e.g. find usages or project search), placing a cursor inside one buffer and editing it, may cause more diagnostics to appear in multiple buffers of the editor (same as inlay hints).

In this logic, we do `lsp_store.document_diagnostic(cursor_buffer, cx)` which is a single buffer, does not seem right — we need to pull diagnostics for every buffer in the editor.

vitallium committed:
Remove redundant `DiagnosticsProvider` trait

vitallium commented on crates/editor/src/editor.rs:
```diff
@@ -14060,6 +14148,46 @@ impl SemanticsProvider for Model<Project> {
     }
 }
 
+impl DiagnosticsProvider for Model<Project> {
```
No, we don't need it. It makes sense to call `lsp_store` directly. Changed that in 9df5721e9d5aec40b2eddd16290f6816967a556b.

vitallium commented on crates/project/src/project.rs:
```diff
@@ -576,6 +577,27 @@ pub const DEFAULT_COMPLETION_CONTEXT: CompletionContext = CompletionContext {
+    pub diagnostics: Option<Vec<lsp::Diagnostic>>,
+}
+
+impl std::default::Default for LspDiagnostics {
```
> it's very scary to use LanguageServerId(0) as a random default value, so let's remove this entirely.

Do you mean to make `server_id` optional? I did it this way because I saw `LanguageServerId(0)` used in other places in the codebase, and I thought this was the default safe value.

SomeoneToIgnore replied:
Maybe I missed something, I've thought that `LanguageServerId(0)` is a valid ID value that can belong to an existing server — please correct me if I'm wrong.

My main point though, is the fact that if we remove `impl Default for LspDiagnostics` entirely, the code compiles just fine (last time I checked, at least) so it's not needed.

vitallium replied:
Ah, yes. I don't remember I added it so, yeah, we can remove it. Removed it in 4df3a5e10043255ab629e235ab41e023e65abc9f Thanks!

vitallium replied:
>Maybe I missed something, I've thought that LanguageServerId(0) is a valid ID value that can belong to an existing server — please correct me if I'm wrong.

I could be wrong as well, but Zed increments this ID per a running LSP starting from 1:

1. Initial value (0) is here https://github.com/vitallium/zed/blob/b906e0f8ef963d1e7c4f7aa6910e055f1cafa9ae/crates/language/src/language_registry.rs#L230 
2. Registering a new LSP increments the ID by 1 here https://github.com/vitallium/zed/blob/b906e0f8ef963d1e7c4f7aa6910e055f1cafa9ae/crates/language/src/language_registry.rs#L973

So a single running LSP will have LanguageServerId equals to 1.

vitallium committed:
remove redundant Default impl for LspDiagnostics

vitallium committed:
log errors when refreshing active diagnostics fails

vitallium committed:
Fix diagnostic req position parameter to use None instead of Some(0)

vitallium committed:
properly handle diagnostic provider identifier

vitallium committed:
improve error handling in diagnostic reporting

vitallium committed:
improve diagnostic tag handling

vitallium committed:
Make diagnostic conversion methods private

vitallium committed:
use LSP adapter disk-based diagnostic sources

vitallium committed:
make diagnostic identifier optional

vitallium committed:
remove clock adv in `move_to_enclosing_bracket`

SomeoneToIgnore reviewed (commented):
Passing by with a few responses and noting that there's https://github.com/zed-industries/zed/pull/23639 trying to do the same thing with a bit different primitives, maybe something interesting can be united.

vitallium commented:
> Passing by with a few responses and noting that there's #23639 trying to do the same thing with a bit different primitives, maybe something interesting can be united.

#23639 looks interesting, thanks for linking it! I probably went into over-engineering mode, and the only things required are in #23639. I am totally fine with giving priority to the linked PR over the somewhat complicated solution in this pull request.

vitallium committed:
Revert "Remove redundant `DiagnosticsProvider` trait"

This reverts commit 53d20495fae6b5edfd078494a12cf4a8e87b5a44.

vitallium committed:
wip

vitallium committed:
wip 2

vitallium committed:
Fix rebase issues

Co-authored-by: Kirill Bulatov <mail4score@gmail.com>

vitallium committed:
Fix a few logical pull diagnostics errors, leave marks for more
improvements

Co-authored-by: Kirill Bulatov <mail4score@gmail.com>

vitallium committed:
Better retain pulled diagnostics

Co-authored-by: Kirill Bulatov <mail4score@gmail.com>

SomeoneToIgnore committed:
Fix the basic test

Co-authored-by: Vitaly Slobodin <vitaly_slobodin@fastmail.com>

SomeoneToIgnore committed:
Work around the hardcoded language server id

SomeoneToIgnore committed:
Merge branch 'main' into vs/lsp-pull-diagnostic-support-with-lsp-command

SomeoneToIgnore committed:
One less TODO(vs)

SomeoneToIgnore committed:
Allow to disable the lsp diagnostics pulling

SomeoneToIgnore committed:
Properly use a strong reference between awaits

SomeoneToIgnore committed:
Pull diagnostics for all buffers

SomeoneToIgnore committed:
Disable pull diagnostics LSP capability if Zed is configured correspondingly

SomeoneToIgnore committed:
Style fixes

SomeoneToIgnore reviewed (commented):
Ok, at this point it seems that we have 

~* one relatively small refactoring to finish~
* to fix a way `pub fn update_diagnostics(` operates: that has to update all related diagnostics there rather than during converting (those are two `TODO(vs)` in the code for)
* consider the sanity of the approach overall and test it more

In particular, I'm interested in `Implementation Considerations` section in https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnostic_refresh

```
Generally the language server specification doesn’t enforce any specific client implementation since those usually depend on how the client UI behaves. However since diagnostics can be provided on a document and workspace level here are some tips:

*    a client should pull actively for the document the users types in.
*    if the server signals inter file dependencies a client should also pull for visible documents to ensure accurate diagnostics. However the pull should happen less frequently.
*    if the server signals workspace pull support a client should also pull for workspace diagnostics. It is recommended for clients to implement partial result progress for the workspace pull to allow servers to keep the request open for a long time. If a server closes a workspace diagnostic pull request the client should re-trigger the request.
```

and whether we are too frequent to query diagnostics for (have larger timeout debounces?).

Otherwise, the code looks nice and quite close to be merged.

SomeoneToIgnore committed:
Merge two pull interface methods into one

SomeoneToIgnore committed:
Merge branch 'main' into vs/lsp-pull-diagnostic-support-with-lsp-command

SomeoneToIgnore committed:
Properly convert pulled diagnostics data

SomeoneToIgnore committed:
Merged in pulled diagnostics more correctly

SomeoneToIgnore committed:
Clippy

SomeoneToIgnore committed:
Fix the tests

SomeoneToIgnore reviewed (commented):
At this point, the initial version seems ready.
We do not have any workspace diagnostics requests (hence, `request_id`s are unaccounted), but can query for diagnostics in all open documents which seems like a good first step.

@vitallium can you test this to be sure it is not broken for your workflows?

SomeoneToIgnore committed:
Merge branch 'main' into vs/lsp-pull-diagnostic-support-with-lsp-command

SomeoneToIgnore committed:
Remove useless `buffer` parameter when merging diagnostics

SomeoneToIgnore committed:
Fix Windows tests, better log URL parsing errors

SomeoneToIgnore committed:
Pull for diagnostics on editor open

SomeoneToIgnore reviewed (approved):
Ok, the initial version is ready, can be disabled with 

```json5
"diagnostics": {
    "lsp_pull_diagnostics_debounce_ms": null
}
```

config and seem to work on @vitallium workflows, so we're ready to merge the initial version, thank you again for bearing through all this.

Both push and pull diagnostics can coexist, defaults for r-a are pushes and the workflow is unchanged.

What's currently missing is https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_diagnostic and https://github.com/zed-industries/zed/pull/19230#pullrequestreview-2868771192 recommendation related to this.

Current impl pulls all open files, handles all related files' diagnostics update and parses the `result_id` from LSP/proto responses.

So, the ground for the workspace diagnostics pulling is ready: we need to store the `result_id`s somewhere and aggregate on pulling the diagnostics.
That will be done separately.

SomeoneToIgnore committed:
Merge branch 'main' into vs/lsp-pull-diagnostic-support-with-lsp-command
